### ISSUE TITLE
[css-lists] CSS counter scope/inheritance is incompatible with HTML ordinals

### ISSUE THREAD

**Issue by MatsPalmgren** [w3c/csswg-drafts#5477]
The rules for counter scope/inheritance in CSS Lists is incompatible with HTML `<ol><ul><li>` ordinals and is thus **not** web-compatible.
https://drafts.csswg.org/css-lists/#nested-counters
https://drafts.csswg.org/css-lists/#inheriting-counters

Gecko has implemented HTML `<ol><ul><li>` ordinals using the built-in `list-item` counter as per the CSS Lists spec. However, we've got many reports that this breaks existing web content to the extent that we need to ship a (non spec-compliant) fix.
Consider this simple HTML example:
```html
<style>
  li::before { content: "(list-item:" counters(list-item,'.') ")"; }
</style>
<ol>
  <li></li>
  <li></li>
  <ol>
    <li></li>
  </ol>
  <li></li>
</ol>
```
According to CSS Lists, the correct rendering is:
![image](https://user-images.githubusercontent.com/4010828/91369856-1b645300-e80d-11ea-8f9a-d99f2c928f61.png)
This is clearly not how HTML lists should work - the last item should have the ordinal 3.
Granted, the above HTML is invalid - [`<ol>`](https://html.spec.whatwg.org/multipage/grouping-content.html#the-ol-element)[`<ul>`](https://html.spec.whatwg.org/multipage/grouping-content.html#the-ul-element) aren't allowed as a direct child of `<ol><ul>`, but this kind of bogus markup is rather common on the web unfortunately. It's also the DOM that `execCommand('indent')` produce (sigh), which affects HTML editors. So telling authors to fix their markup isn't a tenable solution.

The alternative fixes we considered were:
1. change CSS Lists to make all CSS counters behave differently for this case
    Con: breaks existing (author-specified) nested CSS counters of the same name
2. change the behavior for the built-in `list-item` counter only
    Con: makes the built-in `list-item` and other CSS counters behave differently which is a more complicated model to implement, specify, and teach to authors
3. extend CSS Lists to allow either behavior
    (e.g. something like `ol, ul, menu { counter-reset: strict-scope(list-item); }` in the UA sheet)
    Con: a more complicated model to implement, specify, and teach to authors

We concluded that 1 is the least bad solution, so we propose to include a new step in finding the counter scope that searches an element's ancestors first, before falling back to the current steps in the spec that searches previous siblings in reverse DOM order.
This also makes author-defined counters behave more sensible - consider the analogous example:
```html
<style>
  list, item { display: block; }
  list { counter-reset: foo; margin-left: 40px; }
  item { counter-increment: foo; }
  item::before { content: counters(foo,'.'); }
</style>
<list>
  <item></item>
  <item></item>
  <list>
    <item></item>
  </list>
  <item></item>
</list>
```
which currently renders as:
![image](https://user-images.githubusercontent.com/4010828/91371773-3f766300-e812-11ea-98ea-016904e619b8.png)
to instead render as:
![image](https://user-images.githubusercontent.com/4010828/91371817-5c129b00-e812-11ea-9895-cceb50b473ac.png)
which makes more sense IMO.

CC @fantasai @emilio

**Comment by faceless2** [comment]
Slightly nervous about this one. Can I propose a fourth option which might be lower impact?

4. Change the counter rules so that `counter-reset` is applied _after_ `counter-increment` and `counter-set` (it's currently applied before). Then you can cater for the invalid markup with this in a UA sheet.
```css
ol > ol { counter-increment: list-item; counter-reset: list-item }
```
The outer-scoped list-item will be incremented, then a new scope is created with the counter value of 1.

The only impact that would have is on elements that have both a `counter-reset` and `counter-increment` - a pointless combination as the rules are currently defined (and also one that can be searched for to see how often its used). It feels like this approach might have less unintended side-effects.

**Comment by Loirooriol** [comment]
> The outer-scoped list-item will be incremented, then a new scope is created with the counter value of 1.

@faceless2 If `counter-reset: list-item` is applied afterwards, then the inner scope will have value 0. And if counter inheritance prefers the previous sibling over the parent, the next element will still get the value from the inner scope? I don't get how your proposal solves the problem.

**Comment by faceless2** [comment]
I've just worked through it again. I'm wrong and you're correct. Scratch that idea, sorry!

**Comment by tabatkins** [comment]
All right, edits are in; did some editorial cleanup of the algorithm while I was in there, as the structure of the algo didn't make much sense any more.

@MatsPalmgren Mind reviewing to make sure we've caught the right nuances?

**Comment by tabatkins** [comment]
Fun fact: I raised this (on the mailing list) back in 2016! <https://lists.w3.org/Archives/Public/www-style/2016Apr/0364.html>

**Comment by triple-underscore** [comment]
The updated [inherit counters algorithm](https://drafts.csswg.org/css-lists-3/#inherit-counters) by [this commit](https://github.com/w3c/csswg-drafts/commit/ad9a4bc6e6397b6fb66721828e66eb84e73a0c4f) would not produce correct result in certain cases, because it makes elements not inherit a nested counter instantiated by any preceding sibling, due to step 3 of the algorithm:

> Let sibling counters be the CSS counters set of element’s preceding sibling (if it has one), or an empty CSS counters set otherwise.
> For each counter of sibling counters, if element counters does not already contain a counter with the same name, append a copy of counter to element counters.

So, for the following case (no increment), for example:

```
div::before {
    content: "(" counters(foo,'.') ")";
}

<div style="counter-reset: foo">
    <div style="counter-reset: foo"></div>
    <div></div>
</div>
```

According to the algorithm, If my interpretation is correct, the second inner div will only inherit it's parent's counters named `foo`, but ignores the same named nested counter instantiated by the first inner div, and it would produce:

```
(0)
  (0.0)
  (0)
```

But browsers renders this as:

```
(0)
  (0.0)
  (0.0)
```

**Comment by Loirooriol** [comment]
@triple-underscore That was precisely the point of the change. Sure, it's not backwards compatible, but hopefully it will be web compatible. Though it seems that Firefox already got a complaint: https://bugzilla.mozilla.org/show_bug.cgi?id=1672569

**Comment by triple-underscore** [comment]
@Loirooriol
> not backwards compatible, but hopefully it will be web compatible.

Ah, that is what I wanted to know.

Another way to make both backwards compatible and web compatible might be special casing for such mis-parented `<ol>` mentioned in [the first comment](https://github.com/w3c/csswg-drafts/issues/5477#issue-686868744) to have wrapped with “anonymous list item” for the purpose of counters (effectively, the same as option 3 in the first comment), since the author's intention is clear to do so in such cases.

(Or even more, make indeed generate a list item box for such “anonymous list item” for layout purpose, similar to generate anonymous internal table boxes for table layout; of course, it might introduce another compatibility issues.)

**Comment by Loirooriol** [comment]
@MatsPalmgren Can you clarify why Firefox still renders

```html
<style>
  list, item { display: block; }
  list { counter-set: foo; margin-left: 40px; }
  list list { counter-reset: foo; }
  item { counter-increment: foo; }
  item::before { content: counters(foo,'.'); }
</style>
<list>
  <item></item>
  <item></item>
  <list>
    <item></item>
  </list>
  <item></item>
</list>
```
like this
![image](https://user-images.githubusercontent.com/4010828/91371773-3f766300-e812-11ea-98ea-016904e619b8.png)

The top-level `<list>` instantiates the `foo` counter, so why does the last `<item>` fall back to inheriting from the previous sibling? Is this a bug or intentional?

**Comment by MatsPalmgren** [comment]
@Loirooriol [That's intentional](https://searchfox.org/mozilla-central/rev/44e39a47b27313029460bada366c67ebe6d7c57e/layout/base/nsCounterManager.cpp#149-152,158). We only prefer the ancestor's scope when it is an explicit `counter-reset` scope, not an implicit scope created from `counter-set/increment` as in your example (the outer `<list>` only has a `counter-set`, not `counter-reset`).  The reason is that it's the minimal change for making CSS counters compatible with HTML ordinals.

I guess we could include implicit scopes too for consistency perhaps, but that also risks breaking existing content that depends on that behavior. But yeah, it looks like the spec currently doesn't make that distinction, so I guess it's technically a bug in Gecko.  But the intent of our proposal was to only include explicit `counter-reset` scopes, since that was the minimal required change for HTML lists.

I don't feel strongly either way, but I haven't looked at the details of the fallout that change would create... (I can look into that after I'm back from my vacation in case people want to go that route - filed [a bug](https://bugzilla.mozilla.org/show_bug.cgi?id=1718683) for now).

**Comment by Loirooriol** [comment]
@MatsPalmgren Some sites that were broken by the change switched to `counter-set` as a workaround. But implementing the updated spec would break them again. Like https://digital.library.upenn.edu/women/stylesheets/book-2020.css

```css
/* create a page counter for the book */
/* previously used counter-reset in this call */
BODY {
 counter-set: page-counter; 
}
```

It doesn't seem good to do different things for `counter-set` and `counter-reset` when both instantiate a counter, though.

This is getting quite messy. I'm starting to think that a new syntax would have been better (like the recently introduced `reversed()`).

**Comment by Loirooriol** [comment]
Agenda+ to revert the previous resolution, which broke sites and didn't even match the behavior implemented by Firefox, and instead use narrow counter scoped from #9076

**Comment by tabatkins** [comment]
Agreed, [my edits](https://github.com/w3c/csswg-drafts/commit/ad9a4bc6e6397b6fb66721828e66eb84e73a0c4f) do solve the problematic markup given in the OP, but don't solve the closely-related (and totally valid, unlike this issue's) markup given in #9076. So I support reverting and instead introducing a narrow-scoped counter, and using that in HTML's UA stylesheet.

**Comment by Rolf-B** [comment]
Now Firefox follows the spec change, but but not consistently. I declared a nested counter on a span, and it was inherited to a subsequent ::after Element. When I declare the nested counter on a li element, Firefox isolates it, but Chrome inherits it to the next item. So Chrome doesn't respect the Spec change at all, and Firefox does it partially. 

Compare Chrome and Firefox in [this JSFiddle](https://jsfiddle.net/Rolf_b/dtnkmqyu/).

That's really confusing.

**Comment by Loirooriol** [comment]
The last resolution was to revert the previous spec change. So Chrome is right, and if you want to isolate the nested counter you will need #9076

**Comment by josepharhar** [comment]
A WPT change regarding this behavior was proposed here: https://chromium-review.googlesource.com/c/chromium/src/+/5661343

The WPTs here are not yet interoperable: https://wpt.fyi/results/html/semantics/grouping-content/the-li-element?label=master&label=experimental&aligned&q=html%2Fsemantics%2Fgrouping-content

Can someone confirm whether the chromium/webkit or firefox rendering of this is correct? I'm not sure based on the thread of reversing the resolution here, and this doesn't seem to match up perfectly to the example in the OP.
```html
<style>
  body { margin-left: 40px }
  li { list-style-type: decimal }
</style>
<li></li>
<li></li>
<li></li>
<li></li>
<div>
  <li></li>
</div>
<div>
  <div>
    <li></li>
  </div>
</div>
```

Chromium/webkit:
<img width="139" alt="Screenshot 2024-06-28 at 5 50 27 AM" src="https://github.com/w3c/csswg-drafts/assets/7681467/8e2f479c-738d-4a97-8f8d-cb39437054f6">

Firefox:
<img width="139" alt="Screenshot 2024-06-28 at 5 50 41 AM" src="https://github.com/w3c/csswg-drafts/assets/7681467/8bb69e24-e12d-4c59-bf9f-3b6cb5790000">

**Comment by Loirooriol** [comment]
I would say that Firefox is correct, because the `<div>` shouldn't be resetting the `list-item` counter nor anything.
In fact, in Blink I get the same behavior as Firefox if I use `::marker { content: counter(list-item) ". " }`

That should be the case regardless of Mats' original proposal.

**Comment by zp1996** [comment]
@Loirooriol [This test case](https://github.com/web-platform-tests/wpt/blob/master/html/semantics/grouping-content/the-li-element/grouping-li-reftest-list-owner-parent.html) is failed on Firefox. Is Firefox's behavior correct based on the above description?
```html
<style>
  li {
    list-style-type: decimal;
    list-style-position: inside;
  }

  .container {
    padding: 50px;
  }
</style>
<div class="container">
  <li>A</li>
  <li>B</li>
  <div>
    <li>C</li>
    <span>
      <li>D</li>
    </span>
  </div>
  <blockquote>
    <li>E</li>
  </blockquote>
  <li>F</li>
</div>
```
Chromium/Webkit:
![image](https://github.com/user-attachments/assets/6ebcf392-ec55-4872-ac9f-fc907caa618e)
Firefox:
![image](https://github.com/user-attachments/assets/2aead8de-c44c-4793-9409-9f4cdabae89c)

**Comment by Loirooriol** [comment]
Yes, I would again say that Firefox seems right since that's how CSS counters behave, and Blink agrees with `::marker { content: counter(list-item) ". " }`.

**Comment by zp1996** [comment]
@Loirooriol Is the behavior of this test correct in Firefox? 
```html
<style>
  li {
    list-style-type: decimal;
  }
</style>
<div>
  <li></li>
  <li></li>
</div>
<div>
  <li></li>
  <li></li>
</div>
```
actual result:
![image](https://github.com/user-attachments/assets/b1b299f4-434c-4a65-b843-9f2184811218)
IMO, expected result: 
![image](https://github.com/user-attachments/assets/123d49b9-0c4e-468e-9de1-75dbb44ed0ee)

**Comment by Loirooriol** [comment]
Yes, Firefox is correct. The 1st `<li>` instantiates the `list-item` counter, which is propagated to descendants, following siblings, and descendants of following siblings. It's not propagated to following uncles. If you want all the items to share the same instance of the counter, you can use `body { counter-reset: list-item }` (or a `<ol>` wrapper)

**Comment by yisibl** [comment]
Is this the same issue? https://issues.chromium.org/issues/359623686

**Comment by MurakamiShinyu** [comment]
> Is this the same issue? https://issues.chromium.org/issues/359623686

I think so. I encountered this problem when testing a stylesheet that uses `counters()` for nested section numbering, as shown in the following example:

```html
<style>
  section {
    counter-increment: section;
    margin-inline-start: 1em;
  }
  section:first-of-type {
    counter-reset: section;
  }
  section > :is(h1,h2,h3,h4,h5,h6):first-child::before {
    content: counters(section, ".") " ";
  }
</style>
<section>
  <h1>Section Level 1</h1>
  <section>
    <h2>Section Level 2</h2>
  </section>
  <section>
    <h2>Section Level 2, 2nd</h2>
    <section>
      <h3>Section Level 3</h3>
    </section>
    <section>
      <h3>Section Level 3, 2nd</h3>
    </section>
    <section>
      <h3>Section Level 3, 3rd</h3>
    </section>
  </section>
  <section>
    <h2>Section Level 2, 3rd</h2>
  </section>
</section>
<section>
  <h1>Section Level 1, 2nd</h1>
</section>
```

(CodePen: https://codepen.io/MurakamiShinyu/pen/KKjZVbR )

I was surprised that the latest Chrome (126 and later) does not work as expected, while Firefox and Safari work as expected, and I thought it is a bug in Chromium. Then I found this CSS issue.

I believe that reverting the spec change is the right decision, and I hope that all browsers will behave consistently.

**Comment by zp1996** [comment]
According to [recent changes around counter inheritance in the spec](https://drafts.csswg.org/css-lists/#changes-20200709):

> Changed counter inheritance to take from the parent first, and only take from the sibling if it’s a new counter.

But [Example 22 in the spec](https://drafts.csswg.org/css-lists/#example-b1a4542d) conflicts with the spec. In the example, the last `h2`'s counter does not inherit from the parent element, so the last `h2` render like this:
```
B.1 First H2 in H1
``` 

A comparable issue: https://github.com/w3c/csswg-drafts/issues/10491

IMO, we need to update the spec to clarify the behavior.

**Comment by Loirooriol** [comment]
> According to [recent changes around counter inheritance in the spec](https://drafts.csswg.org/css-lists/#changes-20200709)

These changes need to be reverted as resolved in https://github.com/w3c/csswg-drafts/issues/5477#issuecomment-1843956580

**Comment by zp1996** [comment]
As we have reverted the spec, so [the test](https://wpt.fyi/results/css/css-lists/counter-001.html?label=master) results of Firefox and Safari are correct?
```html
<style type="text/css">
#test { counter-reset: c; }
#test span { counter-increment: c; }
#test span::before { content: counter(c, decimal-leading-zero); }
</style>
<p>The following two lines should look the same:</p>
<div id="test">
  <span></span>
  <span></span>
  <span></span>
  <span></span>
  <span></span>
  <span></span>
  <span></span>
  <span></span>
  <span></span>
  <span></span>
  <span></span>
  <span></span>
  <span style="counter-reset: c 98"></span>
  <span></span>
  <span></span>
</div>
```
Chrome inherits from the parent element:
```
01 02 03 04 05 06 07 08 09 10 11 12 99 13 14
```
Firefox and Safari inherit from the sibling element:
```
01 02 03 04 05 06 07 08 09 10 11 12 99 100 101
```

**Comment by Loirooriol** [comment]
Yes, but Firefox seems to have a special behavior for `list-item` (which is odd given that Mats argued against "change the behavior for the built-in `list-item` counter only")

**Comment by emilio** [comment]
Yeah, I changed that in https://bugzilla.mozilla.org/show_bug.cgi?id=1885842 because we kept getting issues like that.
